mtags/src/main/scala-3/scala/meta/internal/pc/ScopeAwareTypePrinter.scala

```diff
@@ -0,0 +1,38 @@
+package scala.meta.internal.pc
+
+import dotty.tools.dotc.core.Flags._
+import dotty.tools.dotc.core.Symbols._
+import dotty.tools.dotc.core.Types._
+import dotty.tools.dotc.printing.RefinedPrinter
+import dotty.tools.dotc.printing.Texts._
+
+class ScopeAwareTypePrinter(indexed: IndexedContext)
+    extends RefinedPrinter(indexed.ctx) {
+
+  private val defaultWidth = 1000
+
+  def typeString(tpw: Type): String = {
+    toText(tpw).mkString(defaultWidth, false)
+  }
+
+  override def toTextPrefix(tp: Type): Text = controlled {
+    tp match {
+      case tp: ThisType =>
+        tp.tref match {
+          case tpe @ TypeRef(NoPrefix, designator) =>
+            val sym =
+              if (designator.isInstanceOf[Symbol])
+                designator.asInstanceOf[Symbol]
+              else tpe.termSymbol
+
+            val text = super.toTextPrefix(tp)
+            if (sym.is(ModuleClass) && indexed.toplevelClashes(sym))
+              Str("_root_.") ~ text
```

Comment:
It might affect hover appearance.
Having `_root_` prefix won't make sence there.

Reply:
the problem is that we will have two different printers and it might be difficult for people to realize which they need. I would rather add a boolean to the other printer like `showRootInConflicts: Boolean`

Reply:
I'm not really sure if it will be better. That would require passing `IndexedContext` in addition to this boolean value.

Actually, creating an additional printer with hacky override was the last thing I wanted to do.

Let's keep it as it is for a while. I think it might be used only in `InferredTypeProvider` and for `implementAbstractMembers`.
And I hope I will find a way to get rid of it :smile:

Reply:
Let's maybe handle it in ShortenedNames similar to how we handle renames? The logic is really connected and it would be better to have it all handled in one place. Especially that it is handled already in Scala 2 this way.